### PR TITLE
Course pages

### PR DESCRIPTION
**Pull request by Sheepsody**
Added a basic system for the courses :
* Can manipulate forms
* Can submit votes for the courses

This PR is the base, but there is still some work to do:
* On the medias for the courses
* On the tags and creation of courses

### CODE REVIEW

**Comment by abocquet** [comment]
image

File: frontend/src/components/Navbar.tsx [hunk 1]
```diff
@@ -26,6 +26,7 @@ const links = [
     { icon: "user", url: "/trombi", label: "Trombi" },
     { icon: "zap", url: "/associations", label: "Associations" },
     { icon: "check-square", url: "/sondages", label: "Sondages" },
+    { icon: "check-square", url: "/cours", label: "Cours" },
```
Comment (abocquet) [inline frontend/src/components/Navbar.tsx:29]: Maybe `book-open` would be more "à propos"

File: frontend/src/components/courses/Home.tsx [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+import Container from "react-bootstrap/Container";
+
+export const CourseHome = ({ course }) => {
+    if (!course) {
+        return "Homepage is loading...";
```
Comment (abocquet) [inline frontend/src/components/courses/Home.tsx:6]: Use the `Loading` component

File: frontend/src/components/courses/List.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                                    </Link>
+                                </Card>
+                            ))}
+                        </Row>
+                        {paginationControl}
+                    </>
+                )}
+        />
+    </Container>
+);
```
Comment (abocquet) [inline frontend/src/components/courses/List.tsx]: Nice usage of `Pagination` 👍

File: frontend/src/components/courses/Router.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                path={match.url + path} // Path is the relative path, match.url makes it absolute
+                component={component}
+                key={path}
+                routeProps={props}
+            />
+        )
+    );
+
+    // Render
+    if (status === "loading") return "Chargement en cours...";
```
Comment (abocquet) [inline frontend/src/components/courses/Router.tsx:41]: Use `Loading` component

File: frontend/src/components/courses/Sidebar.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    if (status === "loading") {
+        return <p>Chargement...</p>;
+        // } else if (error) {
+        //     return <p>Erreur lors du chargement</p>;
+    } else if (course) {
+        return (
+            <Sidebar title={course.name}>
+
+                {/* {user?.isAdmin && */}
```
Comment (abocquet) [inline frontend/src/components/courses/Sidebar.tsx:15]: ?

File: frontend/src/components/courses/Sidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import { UserContext } from "../../services/authService";
+
+export const CourseSidebar = ({ course }) => {
+    let status = "plop";
+    const user: User | null = useContext(UserContext);
+
+    if (status === "loading") {
+        return <p>Chargement...</p>;
+        // } else if (error) {
+        //     return <p>Erreur lors du chargement</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/Sidebar.tsx:11]: Should it be removed ?

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+import { useField, Formik, FormikProps } from "formik";
+import { ToastContext, ToastLevel } from "../../utils/Toast";
+
+export const EvaluateCourse = ({ course }) => {
+    const { data: questions, error, status } = useBetterQuery<Question[]>(
+        ["courses.questions", course.form],
+        api.courses.forms.questions.list,
+    );
+
+    if (status === "loading") return <p>Chargement des cours</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:16]: Use `Loading` component

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+                <PageTitle>Cours</PageTitle>
+                <QuestionsForm questions={questions} course={course} />
+            </Container>
+        );
+    }
+
+    return null;
+};
+
+interface Values {
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:31]: `Values` seems a bit too generic

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    ratings: { [key: number]: number };
+    comments: { [key: number]: string };
+}
+
+
+export const QuestionsForm = ({ questions, course }) => {
+    const newToast = useContext(ToastContext);
+    let [hasVoted, setHasVoted] = useState<boolean>(false);
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:38]: The TS compiler will infer it's a boolean on its own

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        };
+        for (let i in questions) {
+            let question: Question = questions[i];
+            if (question.id) {
+                if (question.category === 'R') {
+                    base.ratings[question.id] = -1;
+                } else if (question.category === 'C') {
+                    base.comments[question.id] = "";
+                } else {
+                    console.log("Got unexpected");
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:53]: Remove `console.log` before merging

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+                initialValues={initRating(questions)}
+                onSubmit={submitAnswers}
+                validate={validate}
+            >
+                {(props: FormikProps<Values>) => (
+                    <Form onSubmit={props.handleSubmit}>
+                        {questions.map((question: Question) => {
+                            if (question.archived) return null;
+
+                            let field: JSX.Element = <p>Error</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:168]: Don't over-type you variables

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+                            let bg = 'light';
+                            if (question.id) if (props.errors[question.id] != undefined) bg = 'danger';
+
+                            return (
+                                <Col md={8} key={question.id}>
+                                    <Card
+                                        key={questions.id}
+                                        className="col-md-4 m-4"
+                                        text={bg === 'light' ? 'dark' : 'white'}
+                                        // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:202]: Why  ?

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+                            let field: JSX.Element = <p>Error</p>;
+                            if (question.category === "R") {
+                                field = <RatingField question={question} id={question.id} name="ratings" label="First Name" {...props} />;
+                            }
+                            else if (question.category === "C") {
+                                field = <CommentField question={question} id={question.id} name="comments" label="First Name" {...props} />;
+                            };
+
+                            let bg = 'light';
+                            if (question.id) if (props.errors[question.id] != undefined) bg = 'danger';
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx]: Maybe use `&&` ?

File: frontend/src/components/courses/evaluations/Evaluate.tsx (hunk 11 above)
Comment (abocquet) [reply frontend/src/components/courses/evaluations/Evaluate.tsx:168]: Use  ternary condition

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+                )
+                }
+            </Formik>
+        </Row>
+
+    )
+};
+
+export const RatingField = ({ question, label, ...props }) => {
+    // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:224]: Why ?

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+                    }
+                </Button>
+            )}
+        </Row>
+    );
+};
+
+
+export const CommentField = ({ question, label, ...props }) => {
+    // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:253]: Why ?

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+        field.value[question.id] = value;
+        helpers.setValue(field.value);
+    };
+
+    if (meta.touched && meta.error) return <p>{meta.error}</p>
+    return (
+        <Form.Control
+            as="textarea"
+            rows={3}
+            // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:266]: Why ?

File: frontend/src/components/courses/evaluations/PaginatedModalComment.tsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+import { QuestionCategory, Question } from "../../../models/courses/question";
+import { StatsQuestion, Comment } from "../../../models/courses/requests";
+import { ColumnChart } from 'react-chartkick';
+import { ToastContext, ToastLevel } from "../../utils/Toast";
+import { Pagination } from "../../utils/Pagination";
+import 'chart.js';
+
+
+export const PaginatedModalComment = ({ question, course }) => {
+    const [show, setShow] = useState<boolean>(false);
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/PaginatedModalComment.tsx:9]: Don't overtype

File: frontend/src/components/courses/evaluations/Stats.tsx [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+    )
+}
+
+const StatsCourse = ({ course }) => {
+    const { data: stats, error, status } = useBetterQuery<StatsQuestion[]>(
+        ["courses.stats", course.id],
+        api.courses.stats,
+    );
+
+    if (status === "loading") return <p>Chargement des cours</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Stats.tsx:61]: Use ternary condition

File: frontend/src/components/courses/evaluations/Stats.tsx [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+                ))}
+            </Row>
+        );
+    }
+
+    return null;
+};
+
+export const PaginatedCardComment = ({ question, course }) => {
+    const [show, setShow] = useState<boolean>(false);
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Stats.tsx]: `show` is never used

File: frontend/src/components/courses/evaluations/Stats.tsx [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    )
+}
+
+const PaginatedComments = ({ course }) => {
+    const { data: questions, error, status } = useBetterQuery<Question[]>(
+        ["courses.forms.question.list", course.form],
+        api.courses.forms.questions.list,
+    );
+
+    if (status === "loading") return <p>Chargement des cours</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Stats.tsx:95]: Use ternary condition

File: frontend/src/components/utils/Toast.tsx [hunk 21]
```diff
@@ -41,7 +42,7 @@ export const ToastProvider: React.FunctionComponent = ({ children }) => {
     );
 };
 
-export const Toast = ({ message: msg, type, flip }) => {
+export const Toast = ({ message: msg, type, flip, delay }) => {
```
Comment (abocquet) [inline frontend/src/components/utils/Toast.tsx:50]: delay is never  used

File: frontend/src/components/utils/forms/DatePickerField.tsx [hunk 22]
```diff
@@ -50,3 +51,71 @@ export function DatePickerField({ label, ...props }: any) {
         </Form.Group>
     );
 }
+
+export function DateTimePickerField( { label, ...props } :any) {
```
Comment (abocquet) [inline frontend/src/components/utils/forms/DatePickerField.tsx]: Sync with @Piwakk on this

File: frontend/src/models/associations/election.ts [hunk 23]
```diff
@@ -2,23 +2,33 @@ import { User } from "../user";
...
-    id: string;
+    id?: string;
     election: Election;
-    choices: Choice[];
+    choices: Choice[] | number[];
 }
 
 export interface Election {
-    id: string;
+    id?: string;
```
Comment (abocquet) [inline frontend/src/models/associations/election.ts:24]: Out of the PR  scope

File: frontend/src/services/api/courses.ts [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+            )
+        ),
+    get: courseId =>
+        unwrap<Course>(
+            apiService.get(
+                `/courses/courses/${courseId}`
+            )
+        ),
+    save: course => {
+        if (!course.id) {
```
Comment (abocquet) [inline frontend/src/services/api/courses.ts:20]: Use ternary condition

File: frontend/src/services/api/courseForms.ts [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+                    `/courses/forms/`,
+                    form,
+                )
+            )
+        )
+    },
+
+    questions: {
+        save: (question: Question) => {
+            console.log("Begore " + question.form)
```
Comment (abocquet) [inline frontend/src/services/api/courseForms.ts]: Remove console.log

File: frontend/src/services/api/courseForms.ts [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+            )
+        ),
+    list: () =>
+        unwrap<PaginatedResponse<Form[]>>(
+            apiService.get(
+                `/courses/forms`
+            )
+        ),
+    save: (form: Form) => {
+        if (form.id)
```
Comment (abocquet) [inline frontend/src/services/api/courseForms.ts:11]: Use ternary condition

File: frontend/src/services/api/courseForms.ts [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+                    form,
+                )
+            )
+        )
+    },
+
+    questions: {
+        save: (question: Question) => {
+            console.log("Begore " + question.form)
+            if (!question.id) {
```
Comment (abocquet) [inline frontend/src/services/api/courseForms.ts:21]: Use ternary condition

File: frontend/src/services/apiService.ts [hunk 28]
```diff
@@ -21,9 +21,16 @@ import { jwt } from "./api/jwt";
 import { users } from "./api/users";
 import { fundings } from "./api/fundings";
 import { profile } from "./api/profile";
+import { courses } from "./api/courses";
 
 const baseApi = "http://localhost:8000/api/v1";
 
+export enum electionActiveStatus {
```
Comment (abocquet) [inline frontend/src/services/apiService.ts:28]: Doesn't  belong  here

Comment (abocquet) [reply frontend/src/services/apiService.ts:28]: Also capitalize `enum` names

File: frontend/src/utils/navbar.tsx [hunk 29]
```diff
@@ -0,0 +1,124 @@
+import React, { useState } from "react";
```
Comment (abocquet) [inline frontend/src/utils/navbar.tsx:1]: The file has moved, do recreate it at its former place

File: backend/courses/models.py [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+        ]
+
+class CourseMedia(models.Model):
+    id = models.AutoField(primary_key=True, unique=True)
+
+    name = models.CharField(max_length=128)
+
+    uploaded_on = models.DateTimeField(auto_now_add=True)
+
+    category = models.CharField(max_length=64)
```
[the remark this answers is not in the document]
Comment (Sheepsody) [reply backend/courses/models.py:80]: Category contains the type of the media
In the future, this could enable us to have both filtering and different behaviours if we embed viewers

File: backend/courses/views/CourseViews.py [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+    def has_voted(self, request, pk=None):
+        current_user = request.user
+        course = self.get_object()
+
+        has_voted = Course.objects.filter(id=course.id, have_voted=current_user).exists()
+
+        return Response({"has_voted": has_voted}, status=status.HTTP_200_OK)
+
+    @action(detail=True, methods=["post"], permission_classes=[IsAuthenticated])
+    def submit(self, request, pk=None):
```
[the remark this answers is not in the document]
Comment (Sheepsody) [reply backend/courses/views/CourseViews.py:40]: Submit is used when the user send the answer to the form associated to that course

File: backend/courses/tests/test_comments.py [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+        "date",
+    ]
+
+    ALL_USERS = ["17admin", "17simple"]
+
+    def endpoint_list(self):
+        return "/courses/comments/"
+
+    def list(self, data=None, courseat="json", content_type="application/json"):
+        return self.post(self.endpoint_list(), data, content_type)
```
[the remark this answers is not in the document]
Comment (Sheepsody) [reply backend/courses/tests/test_comments.py:30]: Changed to query params

File: frontend/src/services/apiService.ts (hunk 28 above)
Comment (Sheepsody) [reply frontend/src/services/apiService.ts:28]: Check with master branch and removed this element

File: frontend/src/utils/navbar.tsx (hunk 29 above)
Comment (Sheepsody) [reply frontend/src/utils/navbar.tsx:1]: Checked again with master branch and find

File: frontend/src/components/courses/List.tsx (hunk 3 above)
Comment (Sheepsody) [reply frontend/src/components/courses/List.tsx]: Thanks !

File: frontend/src/components/courses/Sidebar.tsx (hunk 6 above)
Comment (Sheepsody) [reply frontend/src/components/courses/Sidebar.tsx:11]: Removed, this was old code

File: frontend/src/components/courses/evaluations/Evaluate.tsx (hunk 8 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Evaluate.tsx:31]: Replaced it with EvalSubmission which seems more relevant

File: frontend/src/components/courses/evaluations/Evaluate.tsx (hunk 11 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Evaluate.tsx:168]: I don't see how the ternary condition would improve the code...

File: frontend/src/components/courses/evaluations/Evaluate.tsx (hunk 12 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Evaluate.tsx:202]: Replaced with ternary condition and removed the ignore

File: frontend/src/components/courses/evaluations/Evaluate.tsx (hunk 13 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Evaluate.tsx]: idem

File: frontend/src/components/courses/evaluations/Stats.tsx (hunk 18 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Stats.tsx:61]: There is 3 conditions... Why would it be better ?

File: frontend/src/components/courses/evaluations/Stats.tsx (hunk 19 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Stats.tsx]: Outdated

File: frontend/src/components/courses/evaluations/Stats.tsx (hunk 20 above)
Comment (Sheepsody) [reply frontend/src/components/courses/evaluations/Stats.tsx:95]: idem

File: backend/courses/models.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (abocquet) [reply backend/courses/models.py]: In this case, maybe `mimetype` would be more expressive and consistent with `medias`.  Also see medias to see how it's determined

**Comment by abocquet** [comment]
Got error 
```django.db.utils.DataError: Problem installing fixture 'Portail-des-eleves/backend/courses/fixtures/courses.yaml': Could not load courses.CourseMedia(pk=1): value too long for type character varying(1)
```
when starting launching the fixtures

**Comment by abocquet** [comment]
Also please fix merge issues

**Comment by abocquet** [comment]
image

**Comment by abocquet** [comment]
image

**Comment by abocquet** [comment]
image

Please use French

**Comment by abocquet** [comment]
image

When trying to create a new question

**Comment by Piwakk** [comment]
Hi @Sheepsody and thanks for the hard work!

I am adding some comments on the look and feel of your module — not on the code, for now. I ordered my suggestions by priority: the most important first (in my opinion of course). Feel free to discuss them. @abocquet please comment if you feel so.

As a general principle: Keep it simple!

**Comment by Piwakk** [comment]
**General organisation**

Correct me if I am wrong, but I think this is the flow of your module:
1. The global administrator defines _formulaires_. A _formulaire_ contains questions of a certain type in a certain order.
2. Then, the global administrator creates a _cours_ (with a title and maybe something else) and binds it to a previously created _formulaire_.
3. Finally, the regular users can answer the _formulaire de cours_.

The problem is, when I go to http://localhost:3000/cours/1/ (so I expect to manage a _cours_ or answer a _formulaire de cours_, not to manage the top level _formulaires_), I am given the option to create a new _formulaire_, or to edit the current _formulaire_. These actions actually concern the step 1. and should not be available at the _cours_ level.

So I'd recommend that you create a page dedicated to managing the _formulaires_ (a table could perfectly do the trick, for instance). This page would be accessible from http://localhost:3000/cours (that is the very first page of the module), and give the appropriate links for adding / editing the _formulaires_.

Then, for each _cours_, you'd create a single page _parametres_ or _gerer_, where you could choose the _formulaire_ linked to the _cours_, change the title, etc.

This way it'll be much simpler I think.

**Comment by Piwakk** [comment]
**_Cours_ page**

No need to add an extra page _resultats_. They should be displayed on the “home page” where there is currently the “Home page !!!” message. And I think the `Évaluer` _button_ (it's not a button but it should be) does not have to be in the sidebar, but could be in the main content. But it's a matter of personal taste.

**Comment by Piwakk** [comment]
**Sidebar**

* You should drop entirely the titles (“Formulaires” and “Évaluations”) and add a space instead — everybody will perfectly understand the sidebar items, no need to overload the sidebar. Look at how we do at http://localhost:3000/associations/bde/magasin for instance.
* Please use the same style for the sidebar as in the rest of the site. Use the components `SidebarSpace`, `SidebarSeparator`, `SidebarItem` for instance.
* If something is an action, such as `Créer un formulaire`, then it should be rendered as a button. Otherwise, it's confusing. Again, you can take the Association pages as an example.

**Comment by Piwakk** [comment]
**Évaluer**

I think there is a bug. And you should use Tabler's form styles.

image

**Comment by Piwakk** [comment]
**Lier un formulaire**

Applying a coloured background to a whole card is not something which should be done: too agressive. Instead, use the `CardStatus` component, which will yield a result like on https://preview.tabler.io/cards.html.

I also find it quite weird to create a card for a piece of text and a button. Two choices here:
* Use something else, like a table or a dead-simple radio input.
* Add information to the cards, like the questions used in the _formulaire_. You actually kind of did it in the _Récupérer d’un autre formulaire_ button.

For the change confirmation, prefer using [`window.confirm`](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm), it's much lighter. There are already examples in the code base (it's very simple).

Clicking on _Choisir_ does not update the formulaire, but instead create another _cours_. This must be fixed.

**Comment by Piwakk** [comment]
**Modifier un formulaire**

I am ill at ease with a few things there. I think the best starting point is to look at how [Google](https://forms.google.com) does it:
image

Of course, the drag and drop is not necessary for now, as well as taking the focus into account, but there are a few things which could be easily implemented:
* One question per row, or it feels like each question is independent from the others and they are not just a part of a bigger ensemble (the _formulaire_).
* Do not add a *Valider* button under _each_ question, this is too heavy and a bad user experience (I don't want to click on Valider every time I change a field). A single button to save the state of the whole _formulaire_ is more than enough.
* Also, I think the *Réinitialiser* button is not useful, given that it only resets two parameters. Keep it simple. ;)
* Getting back to the Google example, you see that when creating the _formulaire_, it actually looks like the _real_ form which will be sent to the user. Try to achieve this by using the styles which mimic the “real” ones (see https://preview.tabler.io/form-elements.html)
* After validating a field, I think the green, full border around the card is too much (use the card status as said before). Also, editing the card afterwards does not reset this border.
* Why do the _new_ questions can change their *Catégorie* but not the old ones?
* When adding _two_ (or more) new questions, changing the parameters of the second one will always change the state of the last one.

I also have a few things to say about the *Récupération* Modal:
* No need to start with a default *--Formulaire--* choice — especially when it yields an error when it is selected. Instead, just start with the first available _formulaire_ — or prevent the Modal from appearing if there is none.
* You should be more explicit about the type of questions: just displaying `R` or `C` is confusing, either display the whole word (like `Commentaire` / `Note`) or a symbol (`fe-edit-2` / `fe-star`)

**Comment by Piwakk** [comment]
**Statistiques**
* The modal for the _comment_ questions does not close.
* We are as conservative as possible about adding dependencies. Hence, I am not convinced by the need for react-chart. I feel like something similar and simpler could be achieved with https://react-bootstrap.github.io/components/progress/ for instance, which is a default Bootstrap component.

For instance, see how Apple does it:

image

Or Google:

image

File: backend/courses/models.py [hunk 33]
```diff
@@ -30,7 +30,7 @@
 
 
 class Form(models.Model):
-    id = models.AutoField(unique=True, primary_key=True)
+    id = models.BigAutoField(unique=True, primary_key=True)
```
[the remark this answers is not in the document]
Comment (Sheepsody) [reply backend/courses/models.py:33]: To ensure that is is positive

[the remark this answers is not in the document]
Comment (Piwakk) [reply backend/courses/models.py:33]: I can't find where in the docs it is written that `BigAutoField` make the field positive.
Also, Django will automatic add an `id` field: https://docs.djangoproject.com/en/3.0/topics/db/models/#automatic-primary-key-fields
So this should be removed altogether

[the remark this answers is not in the document]
Comment (Sheepsody) [reply backend/courses/models.py:33]: Removed

**Comment by Sheepsody** [comment]
Hey! Sorry for the delay! I updated everything as you suggested.
Of course, some improvements would increase user experience, but don't you think it is more suitable for another PR ?

File: cli/create_and_populate_database.bash [hunk 34]
```diff
@@ -11,20 +11,22 @@ rm -Rf  repartitions/migrations/*
...
+python3 manage.py makemigrations authentication
+python3 manage.py makemigrations associations
+python3 manage.py makemigrations polls
+python3 manage.py makemigrations repartitions
+python3 manage.py makemigrations subscriptions
+python3 manage.py makemigrations tags
+python3 manage.py makemigrations courses 
 
-python manage.py migrate
+python3 manage.py migrate
```
[the remark this answers is not in the document]
Comment (abocquet) [reply cli/create_and_populate_database.bash:24]: Please don't mark unresolved conversations as resolved

File: frontend/src/components/courses/Sidebar.tsx [hunk 35]
```diff
@@ -0,0 +1,97 @@
...
+import { Loading } from "../utils/Loading";
+import { UserContext } from "../../services/authService";
+import { SidebarSpace, SidebarSeparator } from "../utils/sidebar/Sidebar";
+import { useLocation } from "react-router-dom";
+
+export const MainSidebar = () => {
+  const location = useLocation();
+
+  // Extracting the course from the URL
+  const regCourse = new RegExp("^/cours/([0-9])/");
```
Comment (abocquet) [inline frontend/src/components/courses/Sidebar.tsx:13]: Use url parameters instead https://reacttraining.com/react-router/web/example/url-params

File: frontend/src/components/courses/Router.tsx [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+        path={match.url + path} // Path is the relative path, match.url makes it absolute
+        component={component}
+        key={path}
+        routeProps={props}
+      />
+    )
+  );
+
+  // Render
+  if (status === "loading") return <Loading />;
```
Comment (abocquet) [inline frontend/src/components/courses/Router.tsx:42]: Please systematically use the following form : 
```
return status === "loading" ? <Loading /> : 
    status === "error" ? <Error /> : 
    status === "success" ? <YourComponent>[...]</YourComponent> : 
    null
;
```

Comment (abocquet) [reply frontend/src/components/courses/Router.tsx:42]: Most component don't use this, make sure they do

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 37]
```diff
@@ -0,0 +1,264 @@
...
+          <Form
+            className="d-flex justify-content-center"
+            onSubmit={props.handleSubmit}
+            as={Card}
+          >
+            {questions.map((question) => {
+              if (question.archived) return null;
+
+              let field: JSX.Element = <p>Error</p>;
+              if (question.category === "R") {
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:163]: Use ternary condition

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 38]
```diff
@@ -0,0 +1,264 @@
...
+            </Button>
+          </Form>
+        )}
+      </Formik>
+    </Row>
+  );
+};
+
+export const RatingField = ({ question, label, ...props }) => {
+  // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:219]: Why is this needed ?

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 39]
```diff
@@ -0,0 +1,264 @@
...
+            <span className="text-dark">☆</span>
+          )}
+        </Button>
+      ))}
+    </Row>
+  );
+};
+
+export const CommentField = ({ question, label, ...props }) => {
+  // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:248]: Why is this needed ?

File: frontend/src/components/courses/evaluations/Stats.tsx [hunk 40]
```diff
@@ -0,0 +1,148 @@
...
+  </Col>
+);
+
+const StatsCardQuestion = ({ statsQuestion }) => {
+  let sum = 0;
+  Object.keys(statsQuestion.histogram)
+    .map((e) => parseInt(e))
+    .forEach((key) => {
+      sum += statsQuestion.histogram[key];
+    });
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Stats.tsx:55]: Use `reduce` to calculate sums https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

File: frontend/src/components/courses/evaluations/Stats.tsx [hunk 41]
```diff
@@ -0,0 +1,148 @@
...
+};
+
+const StatsCourse = ({ course }) => {
+  const { data: stats, error, status } = useBetterQuery<StatsQuestion[]>(
+    ["courses.stats", course.id],
+    api.courses.stats
+  );
+
+  if (status === "loading") return <Loading />;
+  if (status === "error") return <p>{`Something went wrong: ${error}`}</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Stats.tsx:88]: Use ternary condition

File: frontend/src/components/courses/evaluations/Stats.tsx [hunk 42]
```diff
@@ -0,0 +1,148 @@
...
+
+const PaginatedComments = ({ course }) => {
+  const { data: questions, error, status } = useBetterQuery<Question[]>(
+    ["courses.forms.question.list", course.form],
+    api.courses.forms.questions.list
+  );
+
+  if (status === "loading") return <Loading />;
+
+  if (status === "error") return <p>{`Something went wrong: ${error}`}</p>;
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Stats.tsx:123]: Use ternary condition

File: frontend/src/components/courses/forms/Edit.tsx [hunk 43]
```diff
@@ -0,0 +1,182 @@
...
+      </Row>
+
+      <br />
+
+      {questions &&
+        questions.map((question, index) => {
+          const key: string = question.id + question.label;
+
+          return (
+            <Row key={"Row-" + key} onClick={() => setTooltipIndex(index)}>
```
Comment (abocquet) [inline frontend/src/components/courses/forms/Edit.tsx:107]: `key` is used only once, you might directly use it (with string interpolation) instead of separating declaration and usage

File: frontend/src/components/courses/forms/QuestionEditor.tsx [hunk 44]
```diff
@@ -0,0 +1,229 @@
...
+        }}
+      </Formik>
+
+      <QuestionFetchModal formId={formId} trigger={showModal} />
+    </>
+  );
+};
+
+const MyRadioField = ({ label, mapping, disabled, ...props }) => {
+  // @ts-ignore
```
Comment (abocquet) [inline frontend/src/components/courses/forms/QuestionEditor.tsx:209]: Why is this needed ?

File: frontend/src/utils/Sidebar.tsx [hunk 45]
```diff
@@ -0,0 +1,35 @@
...
+  return (
+    <>
+      <h2 className="page-title mb-1 mt-5">{title}</h2>
+      {children}
+    </>
+  );
+};
+
+export const Sidebar = ({ title, children, ...rest }) => {
+  return (
```
Comment (abocquet) [inline frontend/src/utils/Sidebar.tsx:29]: No need to use `return` here

File: frontend/src/utils/Sidebar.tsx [hunk 46]
```diff
@@ -0,0 +1,35 @@
...
+      <span className="icon mr-3">
+        <i className={iconClassName} />
+      </span>
+      {children}
+    </Link>
+  );
+};
+
+export const SidebarCategory = ({ title, children }) => {
+  return (
```
Comment (abocquet) [inline frontend/src/utils/Sidebar.tsx:20]: No need to use `return` here

File: frontend/src/utils/Sidebar.tsx [hunk 47]
```diff
@@ -0,0 +1,35 @@
+import React from "react";
+import { Link } from "react-router-dom";
+
+export const SidebarItem = ({ icon, to, children }) => {
+  const iconClassName = "fe fe-" + icon;
+  return (
```
Comment (abocquet) [inline frontend/src/utils/Sidebar.tsx:6]: No need to use `return` here

File: backend/courses/models.py [hunk 48]
```diff
@@ -0,0 +1,174 @@
...
+
+class Question(models.Model):
+
+    # TODO add test
+    class Meta:
+        unique_together = [['form', 'label']]
+
+    id = models.AutoField(primary_key=True, unique=True)
+
+    label = models.CharField(max_length=64)
```
[the remark this answers is not in the document]
Comment (abocquet) [reply backend/courses/models.py:108]: This comment applies to the whole file

File: frontend/src/components/courses/evaluations/Evaluate.tsx [hunk 49]
```diff
@@ -157,30 +155,24 @@ export const QuestionsForm = ({ questions, course }) => {
...
-                  <CommentField
-                    question={question}
-                    id={question.id}
-                    name="comments"
-                    label="First Name"
-                    {...props}
-                  />
-                );
-              }
+              const field = question.archived ? null : question.category ===
```
Comment (abocquet) [inline frontend/src/components/courses/evaluations/Evaluate.tsx:158]: Field becomes an  unnecessary intermediary here, you can directly inject this where you call `{field}`

**Comment by Sheepsody** [comment]
Ready for merge it seems !
